FluidNC/src/Kinematics/ParallelDelta.cpp

```diff
@@ -0,0 +1,363 @@
+
+/*
+  ==================== How it Works ====================================
+  On a delta machine, Grbl axis units are in radians
+  The kinematics converts the cartesian moves in gcode into
+  the radians to move the arms. The Grbl motion planner never sees
+  the actual cartesian values.
+
+  To make the moves straight and smooth on a delta, the cartesian moves
+  are broken into small segments where the non linearity will not be noticed.
+  This is similar to how Grgl draws arcs.
+
+  If you request MPos status it will tell you the position in
+  arm angles. The MPos will report in cartesian values using forward kinematics. 
+  The arm 0 values (angle) are the arms at horizontal.
+
+  Positive angles are below horizontal.
+
+  The machine's Z zero point in the kinematics is parallel to the arm axes.
+  The offset of the Z distance from the arm axes to the end effector joints 
+  at arm angle zero will be printed at startup on the serial port.
+
+  Feedrate in gcode is in the cartesian units. This must be converted to the
+  angles. This is done by calculating the segment move distance and the angle 
+  move distance and applying that ration to the feedrate. 
+
+  FYI: http://forums.trossenrobotics.com/tutorials/introduction-129/delta-robot-kinematics-3276/
+  Better: http://hypertriangle.com/~alex/delta-robot-tutorial/
+
+
+Default configuration
+
+kinematics:
+  ParallelDelta:
+
+  TODO 
+   - Constrain the geometry values to realistic values.
+   - Implement $MI for dynamixel motors.
+
+*/
+
+namespace Kinematics {
+
+    // trigonometric constants to speed up calculations
+    const float sqrt3  = 1.732050807;
+    const float dtr    = M_PI / (float)180.0;  // degrees to radians
+    const float sin120 = sqrt3 / 2.0;
+    const float cos120 = -0.5;
+    const float tan60  = sqrt3;
+    const float sin30  = 0.5;
+    const float tan30  = 1.0 / sqrt3;
+
+    // the geometry of the delta
+    float rf;                                           // radius of the fixed side (length of motor cranks)
+    float re;                                           // radius of end effector side (length of linkages)
+    float f;                                            // sized of fixed side triangel
+    float e;                                            // size of end effector side triangle
+
+    static float last_angle[MAX_N_AXIS]     = { 0.0 };  // A place to save the previous motor angles for distance/feed rate calcs
+    static float last_cartesian[MAX_N_AXIS] = { 0.0 };  // A place to save the previous motor angles for distance/feed rate calcs
+
+    void ParallelDelta::group(Configuration::HandlerBase& handler) {
+        handler.item("crank_mm", rf, 50.0, 500.0);
+        handler.item("base_triangle_mm", f, 20.0, 500.0);
+        handler.item("linkage_mm", re, 20.0, 500.0);
+        handler.item("end_effector_triangle_mm", e, 20.0, 500.0);
+        handler.item("kinematic_segment_len_mm", _kinematic_segment_len_mm, 0.05, 20.0);  //
```

Comment:
ISTM that "kinematic_" is somewhat redundant, so perhaps just "segments_per_mm"?